Add new tool action for shield blocking, replacing `IForgeItem#isShield`

sciwhiz12 opened the pull request:
~~This PR fixes #8054 by re-adding the missing patch for `IForgeItem#isShield` back to `LivingEntiy#getEquipmentSlotForItem`, as noted in the linked issue. This also adds a test mod which adds a custom shield item, to allow testing this PR specifically and the `isShield` feature in general.~~

This PR fixes #8054 by adding a new tool action for shield blocking, `ToolActions#SHIELD_BLOCK`, and adding the necessary patches to make it work correctly. The `isShield` methods are kept for backwards compatibility, deprecated as to be removed in the next major version, and preserved in mods that already use this by redirecting calls of `canPerformAction` with the shield block action to this method.

**Screenshot of the implemented fix using the steps from the linked issue**:
![](https://user-images.githubusercontent.com/21304337/131725804-d406e274-4a6a-476a-8a05-12c92d5a78c7.png)

sciwhiz12 committed:
Re-add missing patch for isShield in LivingEntity

Also adds a test mod which adds a custom shield item, for testing this
PR and the isShield feature.

Fixes #8054

diesieben07 commented on src/test/java/net/minecraftforge/debug/item/CustomShieldTest.java:
```diff
@@ -0,0 +1,57 @@
+package net.minecraftforge.debug.item;
```
Missing license header.

diesieben07 commented on src/test/java/net/minecraftforge/debug/item/CustomShieldTest.java:
```diff
@@ -0,0 +1,57 @@
+import javax.annotation.Nullable;
+
+@Mod("custom_shield_test")
+public class CustomShieldTest {
```
Code formatting is wrong for this file.

Sirttas commented:
I wold suggest moving away from `isShield` and use a tool action.

sciwhiz12 committed:
Add license header

sciwhiz12 committed:
Fix formatting to align with Forge formatting guidelines

sciwhiz12 committed:
Change from isShield to a new ToolAction

sciwhiz12 committed:
Readd isShield to preserve backwards compatibility

sciwhiz12 commented:
Changed the PR and patches to use a new `ToolAction`, `ToolActions.SHIELD_BLOCK`, instead of checking `isShield(LivingEntity)`. The `isShield` methods are deprecated, and `canPerformAction` is modified to return `true` if the action queried is `SHIELD_BLOCK` and the item is a shield through `isShield`, to preserve backwards compatibility. 

I do wonder whether `IForgeItemStack#isShield` should be redirected to `IForgeItem#canPerformAction` too, to preserve compatibility for items which override `canPerformAction` to act as a shield with mods that call `isShield`. Doing this would mean that the `LivingEntity` parameter of `isShield` when passed to the item will always be `null`, though.

KnightMiner commented on src/main/java/net/minecraftforge/common/extensions/IForgeItem.java:
```diff
@@ -524,7 +525,8 @@ default void setDamage(ItemStack stack, int damage)
     {
-        return false;
+        // Temporary patch to keep #isShield working until its removed; change to `return false` once removed
+        return ToolActions.DEFAULT_SHIELD_ACTIONS.contains(toolAction) && isShield(stack, null);
```
Minor, but I would invert the order. An instance comparison/boolean is faster than a hash set lookup.

This fallback backwards compat does feel a bit messy still, but I suppose not much to do about that.

sciwhiz12 replied:
Done.

sciwhiz12 committed:
Permute conditions in canPerformAction compat patch

Most calls of `isShield` will result in an instanceof comparison or
equals check, which is faster than a hash set lookup.

Sirttas commented on src/main/java/net/minecraftforge/common/extensions/IForgeItem.java:
```diff
@@ -688,12 +690,15 @@ default boolean canDisableShield(ItemStack stack, ItemStack shield, LivingEntity
+     * @param stack  the item stack
+     * @param entity the entity holding the item stack
+     * @deprecated To be removed in 1.18. Override {@link #canPerformAction(ItemStack, ToolAction)} and return
+     * {@code true} if the passed in tool action is contained in {@link ToolActions#DEFAULT_SHOVEL_ACTIONS} or is
```
Should be default shield actions instead of default shovle actions.

sciwhiz12 replied:
Done. _(how did I miss this?!)_

sciwhiz12 committed:
Fix referenced actions set

diesieben07 commented on patches/minecraft/net/minecraft/world/item/ShieldItem.java.patch:
```diff
@@ -0,0 +1,14 @@
+    public static DyeColor m_43102_(ItemStack p_43103_) {
+       return DyeColor.m_41053_(p_43103_.m_41698_("BlockEntityTag").m_128451_("Base"));
+    }
++
++   /* ******************** FORGE START ******************** */
++
++   @Override
++   public boolean canPerformAction(ItemStack stack, net.minecraftforge.common.ToolAction toolAction) {
++      return net.minecraftforge.common.ToolActions.DEFAULT_SHIELD_ACTIONS.contains(toolAction);
++   }
```
Why add this patch here instead of doing what `isShield` used to do in `canPerformAction` (or rather, just call `isShield` which already checks this...)?

sciwhiz12 replied:
First, it is consistent with the pre-existing implementation of `ToolAction`. If I look into `SwordItem`, I see `canPerformAction` overriden. If I look into `PickaxeItem`, I see `canPerformAction` overriden. Same with `ShovelItem`, `AxeItem`, `HoeItem`, and `ShearsItem`. Leaving `ShieldItem` out of this consistency will make it the odd one out, and it will seem odd why shields are given that specific treatment.

Second, it will be easier for the next person who comes to remove `isShield` to do so. They only have to do two things now: remove `isShield` from both `IForgeItem` and `IForgeItemStack`, and change the implementation of `IForgeItem#canPerformAction` to `return false`. Doing the former will cause a compile error in the latter, and the comment in the method body will steer them to change it to `return false`. This also avoids changing the patch later.

Third, if we move the check which exists currently in `isShield` to `IForgeItem#canPerformAction`, then any mod currently made which uses `isShield` in their custom shield item will not work, as then `isShield` would never be called. As I've seen Lex state that Forge 1.17 is effectively API-stable, I'd like to avoid breaking `isShield`. Furthermore, `IForgeItem#canPerformAction` already calls `IForgeItem#isShield`.

diesieben07 replied:
Fair enough

sciwhiz12 committed:
Merge latest 1.17.x branch
